src/compat/crypto.rs

```diff
@@ -4,73 +4,113 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+use core::mem::size_of;
+
 use dusk_bls12_381::BlsScalar;
 use dusk_bytes::Serializable;
-use phoenix_core::{Note, PublicKey, ViewKey};
+use phoenix_core::{
+    transaction::{ArchivedTreeLeaf, TreeLeaf},
+    Note, PublicKey,
+};
 
-use alloc::vec::Vec;
+use alloc::{string::ToString, vec::Vec};
 
+use crate::alloc::borrow::ToOwned;
 use crate::{
     key::{self},
     types::{self},
     utils::{self},
     MAX_KEY, MAX_LEN,
 };
 
+const TREE_LEAF_SIZE: usize = size_of::<ArchivedTreeLeaf>();
+
 /// Returns true or false if the note is owned by the index
 /// if its true then nullifier of that note if sent with it
 #[no_mangle]
 pub fn check_note_ownership(args: i32, len: i32) -> i64 {
-    // we just use BalanceArgs again as we don't want to add more cluter types
-    // when the data you want is the same
-    let types::CheckNoteOwnershipArgs { note, seed } =
-        match utils::take_args(args, len) {
-            Some(a) => a,
-            None => return utils::fail(),
-        };
+    // SAFETY: We assume the caller has passed a valid pointer and len as the
+    // function arguments else we might get undefined behavior
+    let args = unsafe { core::slice::from_raw_parts(args as _, len as _) };
 
-    let seed = match utils::sanitize_seed(seed) {
-        Some(s) => s,
-        None => return utils::fail(),
-    };
+    let seed = &args[..64];
+    let leaves: &[u8] = &args[64..];
 
-    let note: Note = match rkyv::from_bytes(&note) {
-        Ok(n) => n,
+    let seed = match seed.try_into() {
+        Ok(s) => s,
         Err(_) => return utils::fail(),
     };
 
-    let mut is_owned: bool = false;
-    let mut nullifier_found = BlsScalar::default();
-    let mut pk_found: Option<PublicKey> = None;
+    let mut leaf_chunk = leaves.chunks_exact(TREE_LEAF_SIZE);
+    let mut last_pos = 0;
 
-    for idx in 0..=MAX_KEY {
-        let idx = idx as u64;
-        let sk = key::derive_sk(&seed, idx);
-        let vk = ViewKey::from(&sk);
+    let mut notes = Vec::new();
+    let mut nullifiers = Vec::new();
+    let mut block_heights = Vec::new();
+    let mut public_spend_keys = Vec::new();
+    let mut view_keys = Vec::with_capacity(MAX_KEY);
+    let mut secret_keys = Vec::with_capacity(MAX_KEY);
 
-        if vk.owns(&note) {
-            let nullifier = note.gen_nullifier(&sk);
+    for idx in 0..MAX_KEY {
+        let idx = idx as u64;
+        let view_key = key::derive_vk(&seed, idx);
+        let sk = key::derive_sk(&seed, idx as _);
+        view_keys.push(view_key);
+        secret_keys.push(sk);
+    }
```

Comment:
As mentioned during our pairing, my suggestion was using arrays, not vectors, since you know the size and you won't need to make them mutable. So this and the loop above can be replaced by:
```rust
    let view_keys: [ViewKey; MAX_KEY] =
        core::array::from_fn(|i| key::derive_vk(&seed, i as _));
    let secret_keys: [SecretKey; MAX_KEY] = 
        core::array::from_fn(|i| key::derive_sk(&seed, i as _));
```

It's more concise and you won't need dynamic allocation for the vectors neither mutable variable since you populate the arrays immediately when you instance them.